src/App/Fossa/FossaAPIV1.hs

```diff
@@ -236,6 +271,87 @@ getLatestBuild apiOpts ProjectRevision{..} = fossaReq $ do
   response <- req GET (buildsEndpoint baseUrl orgId (Locator "custom" projectName (Just projectRevision))) NoReqBody jsonResponse baseOpts
   pure (responseBody response)
 
+---------- Archive build queueing. This Endpoint ensures that after an archive is uploaded, it is scanned.
+
+archiveBuildURL :: Url 'Https -> Url 'Https
+archiveBuildURL baseUrl = baseUrl /: "api" /: "components" /: "build"
+
+archiveBuildUpload ::
+  (Has (Lift IO) sig m, Has Diagnostics sig m) =>
+  ApiOpts ->
+  ArchiveComponents ->
+  m (Maybe C.ByteString)
+archiveBuildUpload apiOpts archiveProjects = runEmpty $
+  fossaReqAllow401 $ do
+    (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+    let opts = "dependency" =: True <> "rawLicenseScan" =: True
+        archiveBuildContext = "Queuing a build for an archive project"
+    -- The response appears to either be "Created" for new builds, or an error message for existing builds.
+    -- Making the actual return value of "Created" essentially worthless.
+    resp <- context archiveBuildContext $ req POST (archiveBuildURL baseUrl) (ReqBodyJson archiveProjects) bsResponse (baseOpts <> opts)
+    pure (responseBody resp)
+
+---------- The signed URL endpoint returns a URL endpoint that can be used to directly upload to an S3 bucket.
+
+signedURLEndpoint :: Url 'Https -> Url 'Https
+signedURLEndpoint baseUrl = baseUrl /: "api" /: "components" /: "signed_url"
+
+getSignedURL ::
+  (Has (Lift IO) sig m, Has Diagnostics sig m) =>
+  ApiOpts ->
+  Text ->
+  Text ->
+  m SignedURL
+getSignedURL apiOpts revision packageName = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+  let opts = "packageSpec" =: packageName <> "revision" =: revision
+      signedURLContext = "Retrieving a signed S3 URL"
+
+  response <- context signedURLContext $ req GET (signedURLEndpoint baseUrl) NoReqBody jsonResponse (baseOpts <> opts)
+  pure (responseBody response)
+
+---------- The archive upload function uploads the file it is given directly to the signed URL it is provided.
+
+archiveUpload ::
+  (Has (Lift IO) sig m, Has Diagnostics sig m) =>
+  SignedURL ->
+  FilePath ->
+  m String
+archiveUpload signedArcURI arcFile = fossaReq $ do
+  let arcURL = URI.mkURI $ signedURL signedArcURI
+
+  case arcURL >>= useHttpsURI of
+    Nothing -> case arcURL of
+      Nothing -> fatalText ("Error attempting to archive upload file: " <> T.pack arcFile <> ". No signed URL supplied")
+      Just resultURL -> fatalText ("Invalid URL: " <> URI.render resultURL)
```

Comment:
I might be misunderstanding these branches, but I think the `Nothing` case here is actually the "invalid URL" branch, and the `Just` branch here is actually successful/fine -- so no error wrt the URL

I think we can use `fromMaybeText` from `Control.Error.Diagnostics` to help us here:
```hs
fromMaybeText :: Has Diagnostics sig m => Text -> Maybe a -> m a
```

like so:
```hs
let arcURL = URI.mkURI $ signedURL signedArcURI

uri <- fromMaybeText ("Invalid URL: " <> signedURL signedArcURI) arcURL
(url, options) <- fromMaybeText ("Invalid HTTPS URI: " <> T.pack (show uri)) (useHttpsURI uri)
```

Reply:
The `Just` branch here is not successful, it signifies what happens if a user supplies a URL, but cannot be converted to a valid `HTTPS` URI by `useHttpsURI`. The Nothing branch is what happens if `URI.mkURI` fails. The reason I split these was to give better messages.

Regardless, your suggestion still works better here so I'm going to use it.